src/storage/buffer_pool_manager/buffer_pool.rs

```diff
@@ -0,0 +1,432 @@
+use self::contracts::{BufferPool, BufferPoolKey, PageReadLock};
+use crate::shared::definitions::{PageId, TableId};
+use dashmap::DashMap;
+use parking_lot::{lock_api::RawRwLock, Mutex, RawRwLock as PLRawRwLock, RwLock};
+use std::{cell::UnsafeCell, ops::Range};
+
+pub mod contracts {
+    use super::*;
+
+    pub struct PageReadLock<'a> {
+        page_table_item_read_lock_raw: &'a PLRawRwLock,
+        frame_read_lock_raw: &'a PLRawRwLock,
+        pub data: &'a [u8],
```

Comment:
Does the client need to hold a lock on both??

Before we needed to since page table item lock -> frame lock. However, in the new flow, we have locks on both anyways. So, thing the client should be touching is data, and we just want to preserve the integrity of data, and so the frame_read_lock should be sufficient